File: jdollarx/src/main/java/org/lab49/jdollarx/custommatchers/CustomMatchers.java
package org.lab49.jdollarx.custommatchers;

import org.lab49.jdollarx.BasicPath;
import org.lab49.jdollarx.InBrowser;
import org.lab49.jdollarx.Path;
import org.lab49.jdollarx.PathOperators;
import org.hamcrest.Description;
import org.hamcrest.Matcher;
import org.hamcrest.TypeSafeMatcher;

import static org.lab49.jdollarx.custommatchers.CustomMatchersUtil.wrap;

public class CustomMatchers {

    public static Matcher<InBrowser> hasElement(final Path el) {
        return new TypeSafeMatcher<InBrowser>() {
            @Override
            public void describeTo(final Description description) {
                description.appendText("browser page contains " + el);
            }

            @Override
            protected void describeMismatchSafely(final InBrowser browser, final
            Description mismatchDescription) {
                mismatchDescription.appendText(wrap(el) + " is absent");
            }

            @Override
            protected boolean matchesSafely(final InBrowser browser) {
                return browser.isPresent(el);
            }
        };
    }

    public static IsPresentNTimes isPresent(int nTimes) {
        return new IsPresentNTimes(nTimes);
    }

    public static IsPresent isPresent() {
       return new IsPresent();
    }

    public static HasElements hasElements(Path path) {
       return new HasElements(path);
    }

    public static Matcher<Path> isPresentIn(final InBrowser browser) {
        return new TypeSafeMatcher<Path>() {
            private Path el;

            @Override
            public void describeTo(final Description description) {
                description.appendText("browser page contains " + el.toString());
            }

            @Override
            protected void describeMismatchSafely(final Path el, final
            Description mismatchDescription) {
                mismatchDescription.appendText(wrap(el) + " is absent");
            }

            @Override
            protected boolean matchesSafely(final Path el) {
                this.el = el;
                return browser.isPresent(el);
            }
        };
    }


    public static Matcher<Path> isPresentNTimes(final Integer n, final InBrowser browser) {
        return new TypeSafeMatcher<Path>() {
            private int foundNTimes;
            private Path el;

            @Override
            public void describeTo(final Description description) {
                description.appendText(String.format("browser page contains %s %d time%s", wrap(el), n, n!=1 ? "s" : ""));
            }

            @Override
            protected void describeMismatchSafely(final Path el, final
            Description mismatchDescription) {
                mismatchDescription.appendText(wrap(el) + " appears " + foundNTimes + " time" + (foundNTimes!=1 ? "s" : ""));
            }

            @Override
            protected boolean matchesSafely(final Path el) {
               this.el = el;
               foundNTimes = browser.numberOfAppearances(el);
                return foundNTimes == n;
            }
        };
    }

    public static Matcher<BasicPath> isDisplayedIn(final InBrowser browser) {
        return new TypeSafeMatcher<BasicPath>() {
            private BasicPath el;

            @Override
            public void describeTo(final Description description) {
                description.appendText(el + " is displayed");
            }

            @Override
            protected void describeMismatchSafely(final BasicPath el, final
            Description mismatchDescription) {
                mismatchDescription.appendText(el + " is not displayed");
            }

            @Override
            protected boolean matchesSafely(final BasicPath el) {
                this.el = el;
                return browser.isDisplayed(el);
            }
        };
    }

    public static Matcher<Path> isSelectedIn(final InBrowser browser) {
        return new TypeSafeMatcher<Path>() {
            private Path el;

            @Override
            public void describeTo(final Description description) {
                description.appendText(el + " is selected");
            }

            @Override
            protected void describeMismatchSafely(final Path el, final
            Description mismatchDescription) {
                mismatchDescription.appendText(el + " is not selected");
            }

            @Override
            protected boolean matchesSafely(final Path el) {
                this.el = el;
                return browser.isSelected(el);
            }
        };
    }

    public static Matcher<Path> isEnabledIn(final InBrowser browser) {
        return new TypeSafeMatcher<Path>() {
            private Path el;

            @Override
            public void describeTo(final Description description) {
                description.appendText(el + " is enabled");
            }

            @Override
            protected void describeMismatchSafely(final Path el, final
            Description mismatchDescription) {
                mismatchDescription.appendText(el + " is not enabled");
            }

            @Override
            protected boolean matchesSafely(final Path el) {
                this.el = el;
                return browser.isEnabled(el);
            }
        };
    }

    public static Matcher<InBrowser> hasNoElement(final Path el) {
        return new TypeSafeMatcher<InBrowser>() {
            @Override
            public void describeTo(final Description description) {
                description.appendText("browser page does not contain " + wrap(el));
            }

            @Override
            protected void describeMismatchSafely(final InBrowser browser, final
            Description mismatchDescription) {
                mismatchDescription.appendText(el.toString() + " is present");
            }

            @Override
            protected boolean matchesSafely(final InBrowser browser) {
                return browser.isPresent(PathOperators.not(el));
            }
        };
    }

    public static Matcher<Path> isAbsentFrom(final InBrowser browser) {
        return new TypeSafeMatcher<Path>() {
            private Path el;

            @Override
            public void describeTo(final Description description) {
                description.appendText("browser page does not contain " + el);
            }

            @Override
            protected void describeMismatchSafely(final Path el, final
            Description mismatchDescription) {
                mismatchDescription.appendText(el.toString() + " is present");
            }

            @Override
            protected boolean matchesSafely(final Path el) {
                this.el = el;
                return browser.isPresent(PathOperators.not(el));
            }
        };
    }

}

File: dollarx/src/main/java/com/github/loyada/dollarx/custommatchers/hamcrest/CustomMatchers.scala
package com.github.loyada.dollarx.custommatchers.hamcrest

import com.github.loyada.dollarx.util.StringUtil
import com.github.loyada.dollarx.{PathOperators, Path, Browser}
import org.hamcrest.Description
import org.hamcrest.Matcher
import org.hamcrest.TypeSafeMatcher
import StringUtil.wrap

object CustomMatchers {

  def hasElement(el: Path): Matcher[Browser] = {
     new TypeSafeMatcher[Browser]() {
      def describeTo(description: Description) {
        description.appendText("browser page contains " + wrap(el))
      }

      protected override def describeMismatchSafely(browser: Browser, mismatchDescription: Description) {
        mismatchDescription.appendText(wrap(el) + " is absent")
      }

      protected def matchesSafely(browser: Browser): Boolean = {
         browser.isPresent(el)
      }
    }
  }

  def isPresent(nTimes: Int): IsPresentNTimes = {
      IsPresentNTimes(nTimes)
  }

  def isPresent: IsPresent = {
     new IsPresent
  }

  def hasElements(path: Path): HasElements = {
      HasElements(path)
  }

  def isPresentIn(browser: Browser): Matcher[Path] = {
     new TypeSafeMatcher[Path]() {
      private var el: Path = null

      def describeTo(description: Description) {
        description.appendText("browser page contains " + wrap(el))
      }

      protected override def describeMismatchSafely(el: Path, mismatchDescription: Description) {
        mismatchDescription.appendText(wrap(el) + " is absent")
      }

      protected def matchesSafely(el: Path): Boolean = {
        this.el = el
         browser.isPresent(el)
      }
    }
  }

  def isDisplayedIn(browser: Browser): Matcher[Path] = {
     new TypeSafeMatcher[Path]() {
      private var el: Path = null

      def describeTo(description: Description) {
        description.appendText(el + " is displayed")
      }

      protected override def describeMismatchSafely(el: Path, mismatchDescription: Description) {
        mismatchDescription.appendText(wrap(el) + " is not displayed")
      }

      protected def matchesSafely(el: Path): Boolean = {
        this.el = el
         browser.isDisplayed(el)
      }
    }
  }

  def isSelectedIn(browser: Browser): Matcher[Path] = {
     new TypeSafeMatcher[Path]() {
      private var el: Path = null

      def describeTo(description: Description) {
        description.appendText(el + " is selected")
      }

      protected override def describeMismatchSafely(el: Path, mismatchDescription: Description) {
        mismatchDescription.appendText(el + " is not selected")
      }

      protected def matchesSafely(el: Path): Boolean = {
        this.el = el
         browser.isSelected(el)
      }
    }
  }

  def isEnabledIn(browser: Browser): Matcher[Path] = {
     new TypeSafeMatcher[Path]() {
      private var el: Path = null

      def describeTo(description: Description) {
        description.appendText(wrap(el) + " is enabled")
      }

      protected override def describeMismatchSafely(el: Path, mismatchDescription: Description) {
        mismatchDescription.appendText(wrap(el) + " is not enabled")
      }

      protected def matchesSafely(el: Path): Boolean = {
        this.el = el
         browser.isEnabled(el)
      }
    }
  }

  def hasNoElement(el: Path): Matcher[Browser] = {
     new TypeSafeMatcher[Browser]() {
      def describeTo(description: Description) {
        description.appendText("browser page does not contain " + StringUtil.wrap(el))
      }

      protected override def describeMismatchSafely(browser: Browser, mismatchDescription: Description) {
        mismatchDescription.appendText(el.toString + " is present")
      }

      protected def matchesSafely(browser: Browser): Boolean = {
         browser.isPresent(PathOperators.not(el))
      }
    }
  }

  def isAbsentFrom(browser: Browser): Matcher[Path] = {
     new TypeSafeMatcher[Path]() {
      private var el: Path = null

      def describeTo(description: Description) {
        description.appendText("browser page does not contain " + wrap(el))
      }

      protected override def describeMismatchSafely(el: Path, mismatchDescription: Description) {
        mismatchDescription.appendText(wrap(el) + " is present")
      }

      protected def matchesSafely(el: Path): Boolean = {
        this.el = el
         browser.isPresent(!el)
      }
    }
  }
}
